### ISSUE TITLE
hhvm.hhbc growing when executing new code

### ISSUE THREAD

**Issue by dpaneda** [facebook/hhvm#7001]
HHVM Version 3.12.1

It seems that the hhvm.hhbc file is growing when executing new code. New code could be in the form of new PHP files, or new eval code. We found this when executing our unit tests, because our mock system (phpunit) makes heavy use of eval code. This behaviour is not reproduced when hhvm.repo.commit is false.

$ cat a.php 

```
<?php

class A {
        public function go () {
                $code = '$var = ' . rand() . ';';
                $eval_code = eval($code);
        }
}

$a = new A();
$a->go();

$ cat php.ini 
hhvm.jit = true
hhvm.repo.commit = true

$ rm ~/.hhvm.hhbc
$ php -c php.ini a.php
$ ls -l ~/.hhvm.hhbc
-rw-r--r-- 1 dpaneda dpaneda 4240384 Apr 19 12:39 /home/dpaneda/.hhvm.hhbc
$ php -c php.ini a.php
$ ls -l ~/.hhvm.hhbc
-rw-r--r-- 1 dpaneda dpaneda 4242432 Apr 19 12:39 /home/dpaneda/.hhvm.hhbc
$ php -c php.ini a.php
$ ls -l ~/.hhvm.hhbc
-rw-r--r-- 1 dpaneda dpaneda 4242432 Apr 19 12:39 /home/dpaneda/.hhvm.hhbc
$ php -c php.ini a.php
$ ls -l ~/.hhvm.hhbc
-rw-r--r-- 1 dpaneda dpaneda 4244480 Apr 19 12:49 /home/dpaneda/.hhvm.hhbc
```
# Expected result

.hhvm.hhbc file not growing

**Comment by javacty** [comment]
http://lamp.baidu.com/2014/10/17/dong-tai-yu-fa-de-xing-neng-wen-ti-fen-xi/

**Comment by jwatzman** [comment]
This is expected behaviour -- that file is a bytecode cache to avoid HHVM regenerating the HHBC every time. I'm fairly certain it's safe to remove, or you can use various flags to avoid saving into it (not recommended though!)

Why do you care if this file grows? If it's just a disk space issue, you can either use the `hhvm.repo.commit` option you indicate (and @paulbiss might have a better option?) or just remove it every so often. But you shouldn't do that outside of unit tests.

**Comment by Orvid** [comment]
It shouldn't be growing at all after the first run unless you are running a different version of HHVM from what generated the repo in the first place.

**Comment by dpaneda** [comment]
I don't consider this directly an issue, but Orvid told me that the file shouldn't be growing and ask me to create an issue about this problem.

I don't know the technical details of this cache, but from a user perspective, the problem is that you create a CI environment based on HHVM, and after a while you realized that this file is growing forever, probably filling the entire partition. While this is indeed not a problem outside the test environment, it is certainly unexpected, and is not very evident that the solution is set to false the hhvm.repo.commit configuration.

**Comment by jwatzman** [comment]
@Orvid @paulbiss don't we still emit eval'd code into the repo? That sounds like what's happening here.

**Comment by Orvid** [comment]
You should be able to set `hhvm.repo.mode = readonly` in your ini to prevent HHVM from emitting eval'd code to the repo. The docs say it's the default value, but could you try setting it explicitly?

**Comment by dpaneda** [comment]
Sorry about the delay! I test with `hhvm.repo.mode = readonly` but I doesn't seem to work, only `hhvm.repo.commit `seems to be able to stop hhbc from growing.

**Comment by aorenste** [comment]
It sounds like the real issue here is to check that hhvm.repo.mode=readonly is working properly.

**Comment by pedigree** [comment]
I hate to necro post this, but it still happens.  Here is a very crude way to replicate the issue

```
<?php
function gen() {
        $r = rand(1000, 100000);
        $random_number_array = range(0, $r);
        shuffle($random_number_array );
        $random_number_array = array_slice($random_number_array ,0, $r);
        return $random_number_array;
}
while (1) {
        $val = var_export(gen(), true);
        // HHVM fails at __set_state, so just use object cast for now
        $val = str_replace('stdClass::__set_state', '(object)', $val);
        @file_put_contents("key.php", '<?php $key='  . $val . ';', LOCK_EX);
        unset($key);
        include("key.php");
        $fs = filesize('.hhvm.hhbc');
        echo "$fs \n";
}
```

and you get this expected result, which keeps growing in size
...
774799360 
774877184 
775680000 
776421376 
777363456 
778162176 
779268096 
779415552 
780832768 
781819904 
...

This will eventually cause disk full issues on code that uses includes for caching data/templates etc

**Comment by fredemmott** [comment]
Sorry, this is definitely a wontfix given that PHP is no longer supported and this kind of dynamic code generation and execution is not supported in hack.

**Comment by pedigree** [comment]
@fredemmott but it is ...

user@hhvm~# hhvm --version
HipHop VM 4.6.0 (rel)
Compiler: 1558464479_241928550
Repo schema: 3b33846e2aa0fe0c7ac9c330960958b68d4a75f1

user@hhvm~# hhvm hh.php 
822972416 
823369728 
823951360 
825290752 
826781696 
828313600 
829517824 
830390272 
830480384 
831152128 
831922176 
832610304 
833368064 

```
<?hh
function gen() {
        $r = rand(1000, 100000);
        $random_number_array = range(0, $r);
        shuffle(&$random_number_array );
        $random_number_array = array_slice($random_number_array ,0, $r);
        return $random_number_array;
}
while (1) {
        $val = var_export(gen(), true);
        // HHVM fails at __set_state, so just use object cast for now
        $val = str_replace('stdClass::__set_state', '(object)', $val);
        file_put_contents("key.hack", '$key='  . $val . '; ' . "\n ", LOCK_EX);
        unset($key);
        include("key.hack");
        $fs = filesize('.hhvm.hhbc');
        echo "$fs \n";
}
```

**Comment by fredemmott** [comment]
The type checker will not recognize calls to code in files generated like that; pseudo main code is also not valid Hack.

Additionally, require/include and related functions will likely be removed once some ongoing work to replace autoloading is complete and well tested
